frontend/packages/console-app/src/components/tour/tour-context.ts

```diff
@@ -0,0 +1,100 @@
+import { createContext, useReducer, Reducer, Dispatch, ReducerAction } from 'react';
+// eslint-disable-next-line @typescript-eslint/ban-ts-ignore
+// @ts-ignore: FIXME missing exports due to out-of-sync @types/react-redux version
+import { useSelector } from 'react-redux';
+import { getFlagsObject } from '@console/internal/reducers/features';
+import { useExtensions } from '@console/plugin-sdk';
+import { RootState } from '@console/internal/redux';
+import { isGuidedTour, GuidedTour } from '@console/plugin-sdk/src/typings/guided-tour';
+import { TourActions } from './const';
+import {
+  getTourLocalStorageForPerspective,
+  filterTourBasedonPermissionAndFlag,
+  setTourCompletionLocalStorageDataForPerspective,
+} from './utils';
+import { TourDataType } from './type';
+
+export const tourReducer = (state: TourState, action: TourActions) => {
+  const { stepNumber } = state;
+  switch (action) {
+    case TourActions.start:
+      return { closeTour: false, startTour: true, completedTour: false, stepNumber: 0 };
+    case TourActions.next:
+      return {
+        ...state,
+        startTour: false,
+        stepNumber: stepNumber + 1,
+      };
+    case TourActions.back:
+      return {
+        ...state,
+        stepNumber: stepNumber - 1,
+      };
+    case TourActions.pause:
+      return {
+        ...state,
+        closeTour: true,
+      };
+    case TourActions.resume:
+      return { ...state, closeTour: false };
+    case TourActions.complete:
+      return { ...state, closeTour: false, completedTour: true, stepNumber: 0 };
+    default:
+      return state;
+  }
+};
+
+type TourReducer = Reducer<TourState, TourActions>;
+
+type TourContextType = {
+  tourState?: TourState;
+  tourDispatch?: Dispatch<ReducerAction<TourReducer>>;
+  tour?: TourDataType;
+  totalSteps?: number;
+  onComplete?: () => void;
+};
+
+export type TourState = {
+  stepNumber: number;
+  startTour: boolean;
+  closeTour: boolean;
+  completedTour: boolean;
+};
+
+export const TourContext = createContext<TourContextType>({});
+
+export const useTourValuesForContext = (): TourContextType => {
+  // declaring a method for the perspective instead of using getActivePerspective
+  // because importing getActivePerspective in this file throws error
+  // Uncaught ReferenceError: Cannot access 'allModels' before initialization and this hook is used in plugin extension for ContextProvider
+  const activePerspective = useSelector(({ UI }: RootState): string => UI.get('activePerspective'));
+  const flags = useSelector(getFlagsObject);
```

Comment:
Fetching all flags is causing re-renders for some reason.
Better to fetch only the flags needed by the current tour in order to minimize these re-renders.